ast/src/parsed/asm.rs

```diff
@@ -399,23 +400,111 @@ impl Machine {
                         MachineStatement::Pil(_, statement) => {
                             Box::new(statement.symbol_definition_names().map(|(s, _)| s))
                         }
-                        MachineStatement::CallSelectors(_, name) => Box::new(once(name)),
-                        MachineStatement::Degree(_, _)
-                        | MachineStatement::Submachine(_, _, _)
+                        MachineStatement::Submachine(_, _, _)
                         | MachineStatement::InstructionDeclaration(_, _, _)
                         | MachineStatement::LinkDeclaration(_, _)
                         | MachineStatement::FunctionDeclaration(_, _, _, _)
                         | MachineStatement::OperationDeclaration(_, _, _, _) => Box::new(empty()),
                     }
-                }),
+                })
+                .chain(self.arguments.names())
+                .chain(self.properties.names()),
         )
     }
 }
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Default, Clone)]
-pub struct MachineArguments {
+pub struct MachineArguments(pub Vec<Param>);
+
+impl MachineArguments {
+    pub fn names(&self) -> impl Iterator<Item = &String> {
+        self.0.iter().map(|p| &p.name)
+    }
+}
+
+impl TryFrom<Vec<Param>> for MachineArguments {
+    type Error = &'static str;
+
+    fn try_from(params: Vec<Param>) -> std::result::Result<Self, Self::Error> {
```

Comment:
What about duplicate names? Where is that caught?

Reply:
didn't really do anything about arguments besides parsing, will be handled in a separate PR